modules/blockbuilder/tenant_store.go

```diff
@@ -141,213 +87,222 @@ func (s *tenantStore) AppendTrace(traceID []byte, tr []byte, ts time.Time) error
+		start = s.startTime
 	}
-	if end > endOfRange || end < start {
+	if end.After(endOfRange) || end.Before(start) {
 		warn = true
-		end = uint32(startSectionTime.Unix())
+		end = s.startTime
 	}
 
 	if warn {
-		dataquality.WarnBlockBuilderOutsideIngestionSlack(s.headBlock.BlockMeta().TenantID)
+		dataquality.WarnBlockBuilderOutsideIngestionSlack(s.tenantID)
 	}
 
 	return start, end
 }
 
-func (s *tenantStore) Flush(ctx context.Context, store tempodb.Writer) error {
-	// TODO - Advance some of this work if possible
-
-	// Cut head block
-	if err := s.cutHeadBlock(true); err != nil {
-		return err
-	}
+type entry struct {
+	id   common.ID
+	hash uint64
+}
 
-	s.blocksMtx.Lock()
-	defer s.blocksMtx.Unlock()
+type chEntry struct {
+	id  common.ID
+	tr  *tempopb.Trace
+	err error
+}
 
-	var (
-		completeBlocks = make([]tempodb.WriteableBlock, len(s.walBlocks))
-		jobErr         = atomic.NewError(nil)
-		wg             = boundedwaitgroup.New(flushConcurrency)
-	)
+type liveTracesIter struct {
+	entries    []entry
+	liveTraces *livetraces.LiveTraces[[]byte]
+	ch         chan []chEntry
+	chBuf      []chEntry
+	cancel     func()
 
-	// Convert WALs to backend blocks
-	for i, block := range s.walBlocks {
-		wg.Add(1)
-		go func(i int, block common.WALBlock) {
-			defer wg.Done()
+	start, end uint64
+}
 
-			completeBlock, err := s.buildWriteableBlock(ctx, block)
-			if err != nil {
-				jobErr.Store(err)
-				return
-			}
+func newLiveTracesIter(liveTraces *livetraces.LiveTraces[[]byte]) *liveTracesIter {
+	// Get the list of all traces sorted by ID
+	tids := make([]entry, 0, len(liveTraces.Traces))
+	for hash, t := range liveTraces.Traces {
+		tids = append(tids, entry{t.ID, hash})
+	}
 
-			err = block.Clear()
-			if err != nil {
-				jobErr.Store(err)
-				return
-			}
+	slices.SortFunc(tids, func(a, b entry) int {
+		return bytes.Compare(a.id, b.id)
+	})
 
-			completeBlocks[i] = completeBlock
-		}(i, block)
-	}
+	ctx, cancel := context.WithCancel(context.Background())
 
-	wg.Wait()
-	if err := jobErr.Load(); err != nil {
-		return err
+	l := &liveTracesIter{
+		entries:    tids,
+		liveTraces: liveTraces,
+		ch:         make(chan []chEntry, 1),
+		cancel:     cancel,
 	}
 
-	// Write all blocks to the store
-	level.Info(s.logger).Log("msg", "writing blocks to storage", "num_blocks", len(completeBlocks))
-	for _, block := range completeBlocks {
-		wg.Add(1)
-		go func(block tempodb.WriteableBlock) {
-			defer wg.Done()
-			level.Info(s.logger).Log("msg", "writing block to storage", "block_id", block.BlockMeta().BlockID.String())
-			if err := store.WriteBlock(ctx, block); err != nil {
-				jobErr.Store(err)
-				return
-			}
+	go l.iter(ctx)
 
-			metricBlockBuilderFlushedBlocks.WithLabelValues(s.tenantID).Inc()
+	return l
+}
 
-			if err := s.wal.LocalBackend().ClearBlock((uuid.UUID)(block.BlockMeta().BlockID), s.tenantID); err != nil {
-				jobErr.Store(err)
+func (i *liveTracesIter) Next(ctx context.Context) (common.ID, *tempopb.Trace, error) {
```

Comment:
I have mixed feelings about the lack of synchronization. It shouldn't happen, but it's possible for both the channel and the buffer to be empty.

Reply:
Can you clarify more the concern? The channel is the synchronization, the iterator won't finish until the channel is closed.